lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +211,244 @@ class Loader {
+    const count = specifiers.length;
+    const namespaces = new Array(count);
+
+    for (let i = 0; i < count; i++) {
+      const specifier = specifiers[i];
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
+
+      namespaces[i] = module.getNamespace();
     }
+
+    return wasArr ?
+      namespaces :
+      namespaces[0];
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
+  /**
+   * Provide the ready-to-use source (output is something node understands).
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {string} url The URL/path of the module to be loaded
+   * @param {Object} context Metadata about the module
+   * @returns {Object}
+   */
+  async load(url, context = {}) {
+    const { 0: defaultLoader, 1: customLoader } = this.#loaders;
+
+    const loader = customLoader ?? defaultLoader;
+    const loaded = await loader(url, context, defaultLoader);
+
+    if (typeof loaded !== 'object') {
+      throw new ERR_INVALID_RETURN_VALUE(
+        'object',
+        'loader load',
+        loaded,
+      );
     }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
+
+    const {
+      format,
+      source,
+    } = loaded;
+
+    if (format === null) {
+      const dataUrl = RegExpPrototypeExec(
+        /^data:([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/,
+        url,
+      );
+
+      throw new ERR_INVALID_MODULE_SPECIFIER(
+        url,
+        dataUrl ? `has an unsupported MIME type "${dataUrl[1]}"` : ''
+      );
     }
 
-    if (typeof preloadCode !== 'string') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'string', 'loader getGlobalPreloadCode', preloadCode);
+    if (typeof format !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'string',
+        'loader resolve',
+        'format',
+        format,
+      );
     }
-    const { compileFunction } = require('vm');
-    const preloadInit = compileFunction(preloadCode, ['getBuiltin'], {
-      filename: '<preload>',
-    });
-    const { NativeModule } = require('internal/bootstrap/loaders');
-
-    preloadInit.call(globalThis, (builtinName) => {
-      if (NativeModule.canBeRequiredByUsers(builtinName)) {
-        return require(builtinName);
+
+    if (
+      source != null &&
+      typeof source !== 'string' &&
+      !(source instanceof SharedArrayBuffer) &&
+      !(source instanceof Uint8Array)
+    ) throw ERR_INVALID_RETURN_PROPERTY_VALUE(
+      'string, SharedArrayBuffer, or Uint8Array',
+      'loader load',
+      'source',
+      source
+    );
+
+    return {
+      format,
+      source,
+    };
+  }
+
+  preload() {
+    if (!this.#globalPreloaders.length) return;
+
+    for (const preloader of this.#globalPreloaders) {
+      if (preloader === null) return;
```

Comment:
Can a preloader be `null`? It seems everywhere there are being added to `this.#globalPreloaders`, we check for truthyness first.
```suggestion
```

Reply:
I believe that was the check before. I vaguely remember reading something saying it can be null, which is some kind of "do nothing" flag.